src/canisters/frontend/ic-asset/src/upload.rs

```diff
@@ -42,20 +47,22 @@ pub async fn upload(
     )
     .await?;
 
-    let operations = assemble_batch_operations(
-        project_assets,
-        canister_assets,
-        AssetDeletionReason::Incompatible,
-    );
-
-    info!(logger, "Committing batch.");
-
-    let args = CommitBatchArguments {
-        batch_id,
-        operations,
-    };
-
-    commit_batch(canister, args).await?;
+    match api_version(canister).await {
+        0 => {
+            let operations = batch_upload::operations::v0::assemble_batch_operations(
+                project_assets,
+                canister_assets,
+                AssetDeletionReason::Incompatible,
+            );
+            info!(logger, "Committing batch.");
+            let args = canister_api::types::batch_upload::v0::CommitBatchArguments {
+                batch_id,
+                operations,
+            };
+            commit_batch(canister, args).await?;
+        }
+        _ => bail!("unsupported API version"),
```

Comment:
We shouldn't need to bail here.  A v1 api canister should be able to accept a v0 CommitBatchArguments, because v0::BatchOperationKind is a subset of v1::BatchOperationKind.
```suggestion
        _ => bail!("unsupported API version"),
```

Reply:
Thanks, I haven't thought about it from a _subset_-perspective, I will keep that in mind.
  
Currently, there is only one version of the API, which is `0`. This PR won't introduce API version `1`. I believe it's ok to `bail` for the unknown version, no?

Reply:
Consider if this code were used to deploy to a canister that had been upgraded to an api version 1 canister.  What should happen?

Reply:
deploying to v1 canister with v0 tooling should succeed, I agree, but what happens when someone wants to deploy to v22 canister with v0 tooling? Are we planning to always change the API in a way that v0 is the subset of the new version?

Reply:
I also feel that in this case, it's ok to expect the users to upgrade the tooling (under the condition the new version of the tooling supports all api versions)

Reply:
as discussed, instead of bailing, we will rely on candid to return error for the incompatible payload